src/typechecker/FStar.TypeChecker.Util.fs

```diff
@@ -1011,26 +1011,102 @@ let check_trivial_precondition env c =
+        match (SS.compress t).n with
+        | Tm_fvar fv -> S.fv_eq_lid fv C.term_lid
+        | _ -> false
+    in
+    let is_t_term_view t =
+        let t = N.unfold_whnf env t in
+        match (SS.compress t).n with
+        | Tm_fvar fv -> S.fv_eq_lid fv C.term_view_lid
+        | _ -> false
+    in
     let is_type t =
         let t = N.unfold_whnf env t in
         match (SS.compress t).n with
         | Tm_type _ -> true
         | _ -> false
     in
-    match (U.unrefine lc.res_typ).n with
-    | Tm_fvar fv
-        when S.fv_eq_lid fv C.bool_lid
-          && is_type t ->
-      let _ = Env.lookup_lid env C.b2t_lid in  //check that we have Prims.b2t in the context
-      let b2t = S.fvar (Ident.set_lid_range C.b2t_lid e.pos) (Delta_constant_at_level 1) None in
-      let lc = bind e.pos env (Some e) lc (None, U.lcomp_of_comp <| S.mk_Total (U.ktype0)) in
-      let e = mk_Tm_app b2t [S.as_arg e] None e.pos in
-      e, lc
+    let res_typ = U.unrefine lc.res_typ in
+    let head, args = U.head_and_args res_typ in
+    if Env.debug env (Options.Other "Coercions") then
+            BU.print3 "(%s) Trying to coerce from (%s) to (%s)\n"
+                    (Range.string_of_range e.pos)
+                    (Print.term_to_string res_typ)
+                    (Print.term_to_string t);
+
+    // checks if `t2 == erased t1`
+    let is_erased t1 t2 =
+        let head, args = U.head_and_args t2 in
+        match (U.un_uinst head).n, args with
+        | Tm_fvar fv, [(x, None)] ->
+            S.fv_eq_lid fv C.erased_lid && U.term_eq x t1
+        | _ -> false
+    in
+
+    match (U.un_uinst head).n, args with
+    | Tm_fvar fv, [] when S.fv_eq_lid fv C.bool_lid && is_type t ->
+        coerce_with env e lc U.ktype0 C.b2t_lid [] []
+
+
+    | Tm_fvar fv, [] when S.fv_eq_lid fv C.term_lid && is_t_term_view t ->
+        coerce_with env e lc S.t_term_view C.inspect [] []
+
+    | Tm_fvar fv, [] when S.fv_eq_lid fv C.term_view_lid && is_t_term t ->
+        coerce_with env e lc S.t_term C.pack [] []
+
+
+    | _ when is_erased t res_typ ->
+        coerce_with env e lc t C.reveal [env.universe_of env t] [S.iarg t]
+
+    | _ when is_erased res_typ t ->
+        coerce_with env e lc t C.hide [env.universe_of env res_typ] [S.iarg res_typ]
+
     | _ ->
       e, lc
 
+let maybe_coerce env (e:term) (t1:typ) (t2:typ) : term * typ =
+    let lc = U.lcomp_of_comp (S.mk_Total t1) in
+    let e, lc = maybe_coerce_lc env e lc t2 in
+    e, lc.res_typ
+
+(* Coerces regardless of expected type if a view exists, useful for matches *)
+(* Returns `None` if no coercion was applied. *)
+let coerce_views (env:Env.env) (e:term) (lc:lcomp) : option<(term * lcomp)> =
+    let rt = lc.res_typ in
+    let rt = U.unrefine rt in
+    let hd, args = U.head_and_args rt in
+    match (SS.compress hd).n, args with
+    | Tm_fvar fv, [] when S.fv_eq_lid fv C.term_lid ->
+        Some <| coerce_with env e lc S.t_term_view C.inspect [] []
+    | _ ->
+        None
+
```

Comment:
I wonder why this function is needed. 

It seems it is because the directionality of type-checking matches isn't enough for the usual coercion insertion mechanism to kick in?

i.e., the type of the scrutinee is computed without considering the types of the pattern. If it were, then the expected type of matching against the view would cause maybe_coerce_lc to kick in ... right?

Reply:
Yeah, I think so. It's the other way round -- the type of scrutinee becomes the expected type of the patterns.